### DISCUSSION TITLE
Planned: PasDoc 0.17.0 release - July 2023

### DISCUSSION THREAD

**Discussion by michaliskambi** [pasdoc/pasdoc#168] [Announcements]
We have lots of [cool good new stuff](https://github.com/pasdoc/pasdoc/blob/master/ChangeLog.md). And it's been a while since the last release :)

Let's target: December 30th (last Friday of 2022!) for the new PasDoc 0.17.0 release.

I will do the release (for platforms that are built by Jenkins and mentioned in https://pasdoc.github.io/ now -- Windows x86_64, Linux x86_64, Linux arm 32-bit (Raspberry Pi), macOS (x86_64)). I'll post it as GitHub release, using our https://github.com/pasdoc/pasdoc/blob/master/ChangeLog.md contents as the release notes.

I need everyone's help to:

- Test everything however you think is appropriate (you're most welcome to report your tests, even when everything went OK, here). In particular run latest PasDoc on your projects and see if it works. As always, you can compile PasDoc yourself or take a development snapshot from https://pasdoc.github.io/DevelopmentSnapshots .

- If you want to build for some additional platform, not mentioned above -- this is welcome, just send me a binary build using `make dist-xxx` (see https://pasdoc.github.io/ReleaseMaking ). (This offer is open to everyone who have contributed something non-trivial to PasDoc in the past, i.e. I need to have some base prior trust in you to take a binary file from you and release it :) ).

- After the release, I'll of course appreciate the help in "spreading the word" and posting about the release wherever you see fit, like all Pascal communities.

For me, everything is already tested and cool, all automated tests pass of course, and also the docs on https://castle-engine.io/apidoc/html/ are built by the very latest PasDoc and everything is cool (including the new anchors with parameters, that support overloaded routines -- in CGE we use PasDoc PHP output + HTML output and they both use proper anchors, this allows me to use AsciiDoctor macro like `cgeref:TCastleSceneCore.PlayAnimation[]` from https://castle-engine.io/viewport_and_scenes and it generates a link https://castle-engine.io/apidoc/html/CastleSceneCore.TCastleSceneCore.html#PlayAnimation-string-TPlayAnimationLooping-boolean-const- ).

This release date is not set in stone or whatever, we're a small project :) -- if there are objections, or if we'll find some critical problem, we can of course delay the release without any problem, let me know.

**Comment by michaliskambi** [comment]
The date is tomorrow and I can already see that I will fail, I had a difficult week due to being sick. 

Moreover, now I really really want to apply https://github.com/pasdoc/pasdoc/pull/169 before release :)

So let me postpone the date +2 weeks, from December 16th to December 30th.

**Comment by Fr0sT-Brutal** [comment]
Then I think we should wait a bit more and reserve Jan 1st ! :)

**Comment by michaliskambi** [comment]
Sorry for falling on this -- busy Christmas times :) 

I still want to make PasDoc release, but I changed the time to a more vague "January 2023" -- let's see can be finish https://github.com/pasdoc/pasdoc/pull/169 and then do a release.

**Comment by PifPof73** [comment]
Hi, and sorry for my longer abstinance ...
I managed to compile actual source on Mac OSX M1 chip (platform aarch64).
There is one warning and 5 notes, which I would go through anyhow.

I try to produce a file for this dist.
Greetings,
PifPof73

**Reply by michaliskambi** [reply]
Thank you, this sounds good, I would happily use PasDoc compiled for macOS on Aarch64.

And I didn't yet manage to make PasDoc release, I drowned in too many tasks in all my projects lately :) I will make a post here when I can promise some real release time.

As for warnings / notes:

What warnings / notes do you see? Currently when compiling on Linux/x86_64 with FPC 3.2.2, I see these:

```
PasDoc_TagManager.pas(1374,35) Note: Call to subroutine "function IndexText(const AText:AnsiString;const AValues:{Open} Array Of AnsiString):LongInt;" marked as inline is not inlined
PasDoc_Languages.pas(366,3) Note: Local variable "NewLanguageTemplate" not used
PasDoc_Main.pas(657,44) Note: Call to subroutine "function IndexText(const AText:AnsiString;const AValues:{Open} Array Of AnsiString):LongInt;" marked as inline is not inlined
```

They are harmless. Indeed I would prefer to deal with them, following my own advise from another project ( https://castle-engine.io/coding_conventions#_fix_warnings_let_the_compiler_help_you_write_reliable_code ), I like to have a clean compiler output (so that I know to pay attention when something is reported).

1. In case of `IndexText` note, I don't think we can do anything (aside from not using `IndexText` at all, or silencing the note). It is part of FPC `StrUtils` unit, and FPC decides to not inline it.

    I silenced these 2 notes, by `{$ifdef FPC} {$push} {$notes off} {$endif}` ... `{$ifdef FPC} {$pop} {$endif}`.
  
2. `NewLanguageTemplate` are unused, but still compiled, for a good reason, we state it in code:

    ```
    { NewLanguageTemplate value is not actually used. We include it just to
      force developers to keep PasDoc_Languages_Template_New_Language.inc
      in compileable state. }
    NewLanguageTemplate: {$I lang\PasDoc_Languages_Template_New_Language.inc}
    ```
  
    I don't understand why, but trying to silence this note fails. That is, I tried adding `{$ifdef FPC} {$push} {$notes off} {$endif}` ... `{$ifdef FPC} {$pop} {$endif}`, even to surround whole unit `PasDoc_Languages`... but the note is still reported.

**Reply by PifPof73** [reply]
Mac OSX Ventura 13.2, M1 chipset, 
Lazarus v2.2.4, fpc 3.2.2

I just realized, that my Lazarus was compiling for x86_64, so I had to recompile the IDE for aarch64, which was not so easy, like it sound ... but now, here we go:

### pasdoc_console:
Free Pascal Compiler version 3.2.2 [2021/05/16] for aarch64
Copyright (c) 1993-2021 by Florian Klaempfl and others
(1002) Target OS: Darwin for AArch64
(3104) Compiling fcllaz.pas
(3104) Compiling lazaruspackageintf.pas
(9009) Assembling lazaruspackageintf
(9009) Assembling fcllaz
(1008) 124 lines compiled, 0.2 sec
(1022) 4 hint(s) issued
Free Pascal Compiler version 3.2.2 [2021/05/16] for aarch64
Copyright (c) 1993-2021 by Florian Klaempfl and others
(1002) Target OS: Darwin for AArch64
(3104) Compiling pasdoc_package.pas
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Gen.pas
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Items.pas
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Types.pas
(9009) Assembling pasdoc_types
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_StringVector.pas
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Serialize.pas
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_StreamUtils.pas
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Utils.pas
(9009) Assembling pasdoc_utils
(9009) Assembling pasdoc_streamutils
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Versions.pas
(9009) Assembling pasdoc_versions
(9009) Assembling pasdoc_serialize
(9009) Assembling pasdoc_stringvector
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_ObjectVector.pas
(9009) Assembling pasdoc_objectvector
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Hashes.pas
/Users/pifpof/projects/pasdoc/source/component/PasDoc_Hashes.pas(195,3) **Note: (6058) Call to subroutine "procedure FillChar(var x:<Formal type>;count:Int64;Value:Byte);" marked as inline is not inlined**
/Users/pifpof/projects/pasdoc/source/component/PasDoc_Hashes.pas(279,3) **Note: (6058) Call to subroutine "procedure FillChar(var x:<Formal type>;count:Int64;Value:Byte);" marked as inline is not inlined**
(9009) Assembling pasdoc_hashes
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_TagManager.pas
/Users/pifpof/projects/pasdoc/source/component/PasDoc_TagManager.pas(1359,35) **Note: (6058) Call to subroutine "function IndexText(const AText:AnsiString;const AValues:{Open} Array Of AnsiString):LongInt;" marked as inline is not inlined**
(9009) Assembling pasdoc_tagmanager
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_SortSettings.pas
(9009) Assembling pasdoc_sortsettings
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_StringPairVector.pas
(9009) Assembling pasdoc_stringpairvector
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Tokenizer.pas
(9009) Assembling pasdoc_tokenizer
(9009) Assembling pasdoc_items
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Languages.pas
/Users/pifpof/projects/pasdoc/source/component/PasDoc_Languages.pas(366,3) **Note: (5025) Local variable "NewLanguageTemplate" not used**
(9009) Assembling pasdoc_languages
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_HierarchyTree.pas
(9009) Assembling pasdoc_hierarchytree
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Aspell.pas
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_ProcessLineTalk.pas
(9009) Assembling pasdoc_processlinetalk
(9009) Assembling pasdoc_aspell
(9009) Assembling pasdoc_gen
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_GenHtml.pas
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Base.pas
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Parser.pas
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Scanner.pas
(9009) Assembling pasdoc_scanner
(9009) Assembling pasdoc_parser
(9009) Assembling pasdoc_base
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/tipue/PasDoc_Tipue.pas
(9009) Assembling pasdoc_tipue
(9009) Assembling pasdoc_genhtml
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_GenHtmlHelp.pas
(9009) Assembling pasdoc_genhtmlhelp
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_GenLatex.pas
(9009) Assembling pasdoc_genlatex
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_GenSimpleXML.pas
(9009) Assembling pasdoc_gensimplexml
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_OptionParser.pas
(9009) Assembling pasdoc_optionparser
(3104) Compiling /Users/pifpof/projects/pasdoc/source/component/PasDoc_Reg.pas
(9009) Assembling pasdoc_reg
(9009) Assembling pasdoc_package
(1008) 51780 lines compiled, 3.6 sec
(1023) 4 note(s) issued
Free Pascal Compiler version 3.2.2 [2021/05/16] for aarch64
Copyright (c) 1993-2021 by Florian Klaempfl and others
(1002) Target OS: Darwin for AArch64
(3104) Compiling pasdoc.dpr
(3104) Compiling PasDoc_Main.pas
/Users/pifpof/projects/pasdoc/source/console/PasDoc_Main.pas(648,8) **Note: (6058) Call to subroutine "function IndexText(const AText:AnsiString;const AValues:{Open} Array Of AnsiString):LongInt;" marked as inline is not inlined**
(9009) Assembling pasdoc_main
(9009) Assembling pasdoc
(9015) Linking /Users/pifpof/projects/pasdoc/source/console/pasdoc
(1008) 804 lines compiled, 0.9 sec
(1023) 1 note(s) issued

### pasdoc_gui
Nachrichten, Warnungen: 1
Warning: other sources path of "pasdoc_gui" contains "/Applications/Lazarus/lcl", which belongs to package "LCLBase"
Free Pascal Compiler version 3.2.2 [2021/05/16] for aarch64
Copyright (c) 1993-2021 by Florian Klaempfl and others
(1002) Target OS: Darwin for AArch64
(3104) Compiling pasdoc_gui.lpr
(9009) Assembling pasdoc_gui
(9022) Compiling resource /Users/pifpof/projects/pasdoc/source/gui/pasdoc_gui.or
(9015) Linking /Users/pifpof/projects/pasdoc/source/gui/pasdoc_gui
(1008) 19 lines compiled, 1.6 sec

**Reply by michaliskambi** [reply]
Thank you!

I see that in addition to warnings I experience on Linux/x86_64, the compilation on Darwin/AArch64 also complains that 2x calls to `FillChar` are not inlined. I guess I'm inclined to just leave these warnings be -- as we cannot do much about them anyway, except hiding them, and they are platform-specific then.

**Comment by michaliskambi** [comment]
Sorry everyone, I failed to do a PasDoc release earlier. My new plan is to do it in July, in 2-3 weeks from now. We have a lot of new stuff, and I want the world to know about it.
